bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/internal/ExtensibleTrustManager.java

```diff
@@ -0,0 +1,199 @@
+        if (linkedTrustManager == null) {
+            logger.trace("No specific trust manager found, falling back to default");
+            defaultTrustManager.checkServerTrusted(chain, authType, sslEngine);
+        } else {
+            linkedTrustManager.checkServerTrusted(chain, authType, sslEngine);
+        }
+    }
+
+    private X509ExtendedTrustManager getLinkedTrustMananger(X509Certificate[] chain) {
+        try {
+            String commonName = getCommonName(chain[0]);
+
+            X509ExtendedTrustManager trustManager = linkedTrustManager.get(commonName);
+
+            if (trustManager != null) {
+                logger.trace("Found trustManager by common name: {}", commonName);
+                return trustManager;
+            }
+
+            logger.trace("Searching trustManager by Subject Alternative Names: {}",
+                    chain[0].getSubjectAlternativeNames());
+            // @formatter:off
+            return chain[0].getSubjectAlternativeNames().stream()
+                    .map(e -> e.get(1))
+                    .map(Object::toString)
+                    .map(linkedTrustManager::get)
+                    .filter(Objects::nonNull)
+                    .findFirst()
+                    .orElse(null);
+            // @formatter:on
+        } catch (CertificateParsingException e) {
+            throw new IllegalStateException("Problem while parsing certificate", e);
+        }
+    }
+
+    private String getCommonName(X509Certificate x509Certificate) {
+        String dn = x509Certificate.getSubjectX500Principal().getName(X500Principal.RFC2253);
+        for (String group : dn.split(",")) {
+            if (group.contains("CN=")) {
+                return group.trim().replace("CN=", "");
+            }
+        }
+        throw new IllegalStateException("No Common Name found");
+    }
+
+    @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)
+    protected void addTlsCertificateProvider(TlsCertificateProvider tlsCertificateProvider) {
+        String hostName = tlsCertificateProvider.getHostName();
+        if (!linkedTrustManager.containsKey(hostName)) {
+            synchronized (linkedTrustManager) {
+                if (!linkedTrustManager.containsKey(hostName)) {
+                    logger.trace("adding implementation of TlsCertificateProvider for {} to ExtensibleTrustManager",
+                            hostName);
+                    linkedTrustManager.put(hostName,
+                            new TlsCertificateTrustManagerAdapter(tlsCertificateProvider).getTrustManager());
+                    return;
+                }
+            }
+        }
+        logger.error("Ignoring implementation of TlsCertificateProvider for {} because of a duplicate", hostName);
+    }
+
+    protected void removeTlsCertificateProvider(TlsCertificateProvider tlsCertificateProvider) {
+        // TODO this might remove the wrong one, should we do a better check here?
```

Comment:
Hm, I assume this needs to be improved (now or later).
If two TlsCertificateProvider are added for the same hostname and the second one is removed again, the first one is not used anymore.

Assume there is no provider for hostname foo.
All providers in the example use the hostname foo.

* add TlsCertificateProvider_1
* TlsCertificateProvider_1 is used for foo
* add TlsCertificateProvider_2
* TlsCertificateProvider_2 is ignored because there is already an entry for foo
* remove TlsCertificateProvider_2
* the entry for foo is removed
* no trust manager for foo anymore but TlsCertificateProvider_1 is still available

Reply:
Can't we keep a map that key is the hostname and the value is a linked list of all candidates for that hostname.
The first entry in the list is the currently used one.

This will allow to improve the remove method:
* if the removed object is not the first entry of the map, just remove it
* if the removed object is equal to the first entry of that map, the linkedTrustManager entry can be removed
* if the list is empty, the entry of that map could be removed
* if there is another object in the list, this one could be used registered as the new linked trust manager